src/SingleInstanceState.js

```diff
@@ -0,0 +1,130 @@
+/**
+ * Copyright (c) 2015-present, Parse, LLC.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @flow
+ */
+
+import * as ObjectStateMutations from './ObjectStateMutations';
+
+type ObjectIdentifier = {
+  className: string;
+  id: string
+};
+
+let objectState: {
+  [className: string]: {
+    [id: string]: State
+  }
+} = {};
+
+export function getState(obj: ObjectIdentifier): ?State {
+  let classData = objectState[obj.className];
+  if (classData) {
+    return classData[obj.id] || null;
+  }
+  return null;
+}
+
+export function initializeState(obj: ObjectIdentifier, initial?: State): State {
+  let state = getState(obj);
+  if (state) {
+    return state;
+  }
+  if (!objectState[obj.className]) {
+    objectState[obj.className] = {};
+  }
+  if (!initial) {
+    initial = ObjectStateMutations.defaultState();
+  }
+  state = objectState[obj.className][obj.id] = initial;
+  return state;
+}
+
+export function removeState(obj: ObjectIdentifier): ?State {
+  let state = getState(obj);
+  if (state === null) {
+    return null;
+  }
+  delete objectState[obj.className][obj.id];
+  return state;
+}
+
+export function getServerData(obj: ObjectIdentifier): AttributeMap {
+  let state = getState(obj);
+  if (state) {
+    return state.serverData;
```

Comment:
Do we want to be consistent in the default return value for all these functions? 
`return state.serverData || {};` 
like we did in `getState`?